File: apps/web/src/lib/world-coords.ts
/**
 * ⭐⭐ DÜNYA SAYFASININ KOORDİNAT ÖNCELİĞİ — ve «Kendi diyarıma dön» düğmesinin kararı.
 *
 * ⚠️⚠️ **CANLI HATA (kullanıcı, 2026-08-16).** Düğme yalnız `setSel(null)` yapıyordu ve derin
 * bağlantıyla açılmış sayfada YANLIŞ yere gidiyordu: öncelik zinciri `sel → adres → ev` diye
 * ilerliyor, yani seçim boşalınca sıra **eve değil adrese** düşüyordu. Casusluk raporundan
 * Dünya'ya geçen oyuncu düğmeye basınca kendi diyarına değil raporun diyarına dönüyordu.
 *
 * Doğru davranış iki parçalı ve ikisi de gerekli: seçimi bırak **ve** adresi temizle.
 *
 * ⚠️ Karar burada **saf bir fonksiyonda** yaşıyor çünkü projede tarayıcı testi altyapısı yok
 * (`jsdom`/`testing-library` bağımlılığı yok). `deepLinkAction` (`deep-link.ts`), `tipReduce`
 * (`Tooltip.tsx`) ve `placePopover` (`Popover.tsx`) ile aynı gerekçe: bileşenin içinde
 * kalsaydı ancak elle tıklayarak denenebilirdi — nitekim bu hata tam da öyle kaçmıştı.
 */

export interface Coords {
  k: number;
  d: number;
}

/**
 * Ekranda gösterilecek diyar.
 *
 * Öncelik: **elle seçim → adres → aktif şehir → 1:1**.
 *
 * ⚠️ Adres, evin ÖNÜNDE olmak zorunda: derin bağlantıyla («Dünyada Bul», casusluk raporu,
 * paylaşılan adres) gelen oyuncu ilk açılışta hedefi görmeli, kendi diyarını değil.
 * Hatanın kaynağı bu sıra DEĞİLDİ — sıra doğru; eksik olan, düğmenin adresi de bırakmasıydı.
 */
export function visibleCoords(
  sel: Coords | null, fromUrl: Coords | null, home: Coords | null,
): Coords {
  return {
    k: sel?.k ?? fromUrl?.k ?? home?.k ?? 1,
    d: sel?.d ?? fromUrl?.d ?? home?.d ?? 1,
  };
}

/** «Kendi diyarıma dön» düğmesine basılınca ne olmalı. */
export interface HomeAction {
  /**
   * Yeni seçim — daima `null`.
   *
   * ⚠️ Eve ait koordinatı `sel`e YAZMIYORUZ, bilerek: `sel = null` "sabitlemeyi bırak, aktif
   * şehri izle" demek. Oyuncu sonradan şehir değiştirirse görünüm onu takip etmeli; eve ait
   * koordinatı sabitleseydik takip ölürdü ve adres çubuğu da eski koordinatta kalırdı.
   */
  sel: null;
  /** Adresteki `/world/:k/:d` temizlenmeli mi (`/world`e gidilsin mi)? */
  clearUrl: boolean;
}

export function homeAction(fromUrl: Coords | null): HomeAction {
  return { sel: null, clearUrl: fromUrl != null };
}

File: apps/mobile/lib/core/world_coords.dart
/// ⭐⭐ DÜNYA EKRANININ KOORDİNAT ÖNCELİĞİ — hangi diyar gösterilecek.
///
/// Web'deki `lib/world-coords.ts` ile **aynı kural**, aynı gerekçe.
///
/// ⚠️⚠️ Bu kural web'de **canlı bir hataya** yol açtı (kullanıcı, 2026-08-16): «Kendi diyarıma
/// dön» düğmesi yalnız seçimi boşaltıyordu ve derin bağlantıyla açılmış sayfada YANLIŞ yere
/// gidiyordu — zincir `seçim → adres → ev` diye ilerlediği için seçim boşalınca sıra **eve
/// değil adrese** düşüyordu. Casusluk raporundan Dünya'ya geçen oyuncu düğmeye basınca kendi
/// diyarına değil raporun diyarına dönüyordu. Doğru davranış iki parçalı: seçimi bırak **ve**
/// adresi temizle.
///
/// ⚠️ Karar burada **saf bir fonksiyonda** çünkü bileşenin içinde kalsaydı ancak elle
/// tıklayarak denenebilirdi — hata tam da öyle kaçmıştı (`routing_rules.dart`, `train_rules.dart`
/// ile aynı desen).
library;

typedef MwRealm = ({int k, int d});

/// Kıta 1..10, diyar 1..500 (§13.16).
const int kMaxContinent = 10;
const int kMaxRealm = 500;

/// Ekranda gösterilecek diyar.
///
/// Öncelik: **elle seçim → adres → aktif şehir → 1:1**.
///
/// ⚠️ Adres, evin ÖNÜNDE olmak zorunda: derin bağlantıyla («Dünyada Bul», casusluk raporu,
/// bildirim) gelen oyuncu ilk açılışta hedefi görmeli, kendi diyarını değil.
///
/// ⚠️ `k` ve `d` **ayrı ayrı** çözülüyor, çift olarak değil — web'de de öyle. Oyuncu yalnız
/// kıtayı değiştirdiğinde diyar numarası yerinde kalsın diye.
MwRealm visibleCoords(MwRealm? sel, MwRealm? fromUrl, MwRealm? home) => (
  k: sel?.k ?? fromUrl?.k ?? home?.k ?? 1,
  d: sel?.d ?? fromUrl?.d ?? home?.d ?? 1,
);

/// «Kendi diyarıma dön» düğmesine basılınca ne olmalı.
///
/// ⚠️ Eve ait koordinat `sel`e **yazılmıyor**, bilerek: `sel = null` "sabitlemeyi bırak, aktif
/// şehri izle" demek. Oyuncu sonradan şehir değiştirirse görünüm onu takip etmeli; koordinatı
/// sabitleseydik takip ölürdü.
typedef MwHomeAction = ({MwRealm? sel, bool clearUrl});

MwHomeAction homeAction(MwRealm? fromUrl) =>
    (sel: null, clearUrl: fromUrl != null);

/// Adresten gelen koordinat — geçersizse `null`.
///
/// ⚠️ Kelepçeleme ATMA DEĞİL: `11:900` gibi bir adres 10:500'e kırpılıyor, reddedilmiyor.
/// Reddetseydik bozuk bir bağlantı oyuncuyu sessizce kendi diyarına atardı ve o "bağlantı
/// çalışmadı" diye okunurdu; kırpma en azından yakın bir yere götürüyor.
MwRealm? realmFromPath(String? k, String? d) {
  final ki = int.tryParse(k ?? '');
  final di = int.tryParse(d ?? '');
  if (ki == null || di == null || ki < 1 || di < 1) return null;
  return (k: ki.clamp(1, kMaxContinent), d: di.clamp(1, kMaxRealm));
}
